Add threading api based on luasdl2.

jgmdev opened the pull request:
Implemented support for threading and channels by porting luasdl2 related code to lite-xl:
* https://github.com/Tangent128/luasdl2/wiki/Sdl-category-thread
* https://github.com/Tangent128/luasdl2/wiki/Sdl-category-channels

Two new namespaces are exposed and two new objects briefly described below:

```lua
----------------------------------------------------
thread.create(name, function_or_file_path, ...optional_args_if_func) -> Thread
----------------------------------------------------
Thread:getId() -> integer
Thread:getName() -> string
Thread:wait()

----------------------------------------------------
channel.get(name) -> Channel
----------------------------------------------------
Channel:first() -> any
Channel:last() -> any
Channel:push(value) -> bool, string
Channel:supply(value) -> bool, string
Channel:clear()
Channel:pop()
Channel:wait() -> any
```
This would allow to write code that is non blocking with channels based communication, eg:

```lua
local outsidechan = channel.get("mychanreader")
local outsidechan2 = channel.get("mychanwriter")

local mythread = thread.create("mythread", function()
  local insidechan = channel.get("mychanreader")                                                                                                                            
  local insidechan2 = channel.get("mychanwriter")                                                                                                                          
                                                                                                                                                              
  while true  do                                                                                                                                              
    system.sleep(1)                                                                                                                                           
    local value = insidechan:wait()                                                                                                                                 
    if type(value) == "table" then                                                                                                                            
      print (value[1])                                                                                                                                        
    end                                                                                                                                                       
    print(value)                                                                                                                                              
    insidechan:pop()                                                                                                                                                
                                                                                                                                              
    insidechan2:push(value)                                                                                                                                             
  end                                                                                                                                                         
end)                                                                                                                                                          
                                                                                                                                                              
outsidechan:push("some")
outsidechan:push("data")
outsidechan:push("test")
outsidechan:push({1, 2})
```

adamharrison commented:
I'm somewhat against threading in general in lite; it tends to really complicate code, and brings the need for mutexes, and other synchronization primitives.

As it stands, what problem are we trying to solve with this? Ideally we could just use non-blocking APIs (like we do with the process API, or the dirmonitor stuff).

If there's something we can't solve with the existing APIs, I'm generally more for it; what specifically are you going for here?

jgmdev commented:
> I'm somewhat against threading in general in lite; it tends to really complicate code, and brings the need for mutexes, and other synchronization primitives.

This implementation creates a separate lua state so it doesn't allows use of main lite-xl lua state globals, instead the plugin developer would rely on the use of channels to interchange data between the main thread and additional created threads. The channels internally handle mutexes/locking/unlocking making it pretty easy to communicate with your threads or between threads.

> As it stands, what problem are we trying to solve with this? Ideally we could just use non-blocking APIs (like we do with the process API, or the dirmonitor stuff), or lua coroutines, like we do with the `core.add_thread` stuff.

The idea is to provide additional flexibility and better response times not provided by coroutines alone for time consuming tasks where yielding just makes the process slow or the gui stall. 

One example that comes to mind, if I want to scan a 100,000 files directory using `system.list_dir`, using coroutines I would need to yield every certain time which would end on a slower scan of the directory due to the yield interruptions been done to reduce GUI stalls. The scan would get even slower when it needs to wait in a queue for other co-routines to yield too. 

Using a thread would reduce the amount of time for a task to complete which would otherwise take longer using the co-routine method (`core.add_thread`) due to all interruptions, at the same time threads will not cause the gui to feel sluggish or stall since they will not interfere with time that is better spent rendering.

The goal is not to replace coroutines but to complement them in a way that allows more time for the rendering of the gui when in need of running intensive tasks. As an example one may use a coroutine to handle the communication with threads using channels which is a not a time consuming task, while the threads take care of the more time consuming parts. So the idea of using threads only when really needed should be promoted.

> If there's something we can't solve with the existing APIs, I'm generally more for it; what specifically are you going for here?

In summary:
* Provide a mechanism to create 100% non-blocking threads
* Threads don't block other threads and don't need to yield
* Threads don't affect the rendering engine since they are fully separate from main thread resulting in smoother gui
* Easy communication with threads employing channels (no need to worry about mutexes (lock/unlock))
* Expand furthermore the possibilities of the plugins that can be developed for lite-xl (one never knows what a talented contributor can come up to)

Functionality or plugins that would benefit:
* Periodically checking files (eg: TodoTreeView, with co-routines makes gui stall)
* File search and replace (not having to yield would result in faster results for large projects)
* File indexing (core:find-file a thread can be used for instantaneous indexation)
* etc...

adamharrison commented:
I guess I just worry that it's a lot of code for use cases that may come in future; right now, lite is only about 2500 SLOC in C; and this is adding ~700. So we'd be adding 30% extra editor for a use case that currently, I don't think anyone has reported as an issue, yet. (not to say these issues don't exist! It just doesn't seem like people are complaining about things that can be solved with threading at the moment).

Like, say we have a directory tree of 100,000 files. How long does that currently take when we do yield as we do now? I thought it was pretty quick as-is. Generally directory listings are disk-bound, not CPU bound, anyway.

I do understand the benefits of threading in general, of course. I guess I'm not super opposed, but I'd like to hear other commenters.

jgmdev commented:
> Like, say we have a directory tree of 100,000 files. How long does that currently take when we do yield as we do now? 

The yielding can trigger other intensive tasks that can slow the scan, in my testing scanning a large project can take more than 6 seconds while also causing UI stalls when triggering tasks that interfere with a co-routine (for example doing a fuzzy_match on a list of files while searching for more files in a coroutine). With no interruptions time is reduced to 1 second or less on my SSD.

> I do understand the benefits of threading in general, of course. I guess I'm not super opposed, but I'd like to hear other commenters.

Yeah, would be nice to also know others thought!

Guldoman commented:
I understand that threading might be needed for some plugins and this PR seems to provide the functionality in a simple and clear way, but yeah this is a lot of code for a functionality that we don't currently _need_ in core.

Skimming through the code this seems to use SDL threading functions, if we were to expose them to the Lua side, would it be possible to build this as an external plugin?

jgmdev commented:
> I understand that threading might be needed for some plugins and this PR seems to provide the functionality in a simple and clear way,

Yes, luasdl2 code is pretty nice and clean!

> but yeah this is a lot of code for a functionality that we don't currently _need_ in core.

At least two parts of core come to mind that can make use of this, as I wrote above the threads can be used for `core:find-file` (eg: https://github.com/lite-xl/lite-xl-plugins/pull/29) and when in need of scanning a project files for the benefits already mentioned above.

> Skimming through the code this seems to use SDL threading functions, if we were to expose them to the Lua side, would it be possible to build this as an external plugin?

You mean exposing the raw SDL threading functions to write the functionality as lua code or writing a native C plugin using the support that @takase1121 exposed? 

* If the first, I'm not sure how easy or possible would be creating a new lua state from user side which would be required for running a separate thread if all that was exposed where the SDL threading functions. My guess is that it may result in a more complicated setup for threading.
* If the latter, I haven't familiarized my self with writing a native plugin and if all the required lua API functionality is exposed for doing this.

Guldoman commented:
> At least two parts of core come to mind that can make use of this

What I meant was that there's nothing that _really needs_ to be threaded, as the performance in most situations is not bad.

> You mean exposing the raw SDL threading functions to write the functionality as lua code or writing a native C plugin using the support that @takase1121 exposed?

Either one, really. A native plugin could be more complex to support tho.

> * If the first, I'm not sure how easy or possible would be creating a new lua state from user side which would be required for running a separate thread if all that was exposed where the SDL threading functions. My guess is that it may result in a more complicated setup for threading.

I don't know much about Lua-C interaction, so I can't comment on that.

jgmdev commented:
Ok, after the given feedback I removed the functionality from lite-xl core code and added it as separate modules that are located in the `modules/` directory with its own `modules/meson.build` called from the root `meson.build` and are installed to `lite_data_dir/native` they should be loadable by doing a `require "native.thread"` but that path doesn't looks nice... so not sure where to place them, suggestions are welcome, also haven't tested if they are properly loaded by the `f_load_native_plugin` lite-xl function, will come back after weekend to better test this.

Suggestions for the file structure to place the extra lua C modules and where to install them is highly desired :smiley:

jgmdev commented:
After some testing got the module working and made other changes.

### Summary:

* The thread module sources are now stored in `modules/thread/`
* Every future lua C module code should be stored on `modules/{modname}/`
* `modules/meson.build` will contain the recipes to build and install the modules
* modules are installed into `lite_datadir + "/modules/modname.{lib_extension}"`
* All modules on `lite_datadir + "/modules/"` will be optionally loadable using @takase1121 `f_load_native_plugin`
* `lite_datadir + "/modules/"` was added to `package.cpath` so they are easily loadable by doing `modname = require "modname"`
* This method of adding modules will not touch lite-xl core code for extra native functionality that could be contributed in the future like `network support`
* Exposed `api_load_libs` so the thread module can load lite-xl libraries like `system` for functions that can be used on the threads like `system.list_dir()`

### How the thread module looks now:

```lua
----------------------------------------------------
-- thread library
----------------------------------------------------
local thread = require "thread"
thread.create(name, function_or_file_path, ...optional_args_if_func) -> Thread
thread.getChannel(name) -> Channel
----------------------------------------------------
-- Thread Object
----------------------------------------------------
Thread:getId() -> integer
Thread:getName() -> string
Thread:wait()
----------------------------------------------------
-- Channel Object
----------------------------------------------------
Channel:first() -> any
Channel:last() -> any
Channel:push(value) -> bool, string
Channel:supply(value) -> bool, string
Channel:clear()
Channel:pop()
Channel:wait() -> any
```
Updated sample code:

```lua
local threads = require "thread"
local outsidechan = threads.getChannel("mychanreader")
local outsidechan2 = threads.getChannel("mychanwriter")

local mythread = threads.create("mythread", function()
  local insidechan = thread.getChannel("mychanreader")                                                                                                                            
  local insidechan2 = thread.getChannel("mychanwriter")                                                                                                                          
                                                                                                                                                              
  while true  do                                                                                                                                              
    system.sleep(1)                                                                                                                                           
    local value = insidechan:wait()                                                                                                                                 
    if type(value) == "table" then                                                                                                                            
      print (value[1])                                                                                                                                        
    end                                                                                                                                                       
    print(value)                                                                                                                                              
    insidechan:pop()                                                                                                                                                
                                                                                                                                              
    insidechan2:push(value)                                                                                                                                             
  end                                                                                                                                                         
end)                                                                                                                                                          
                                                                                                                                                              
outsidechan:push("some")
outsidechan:push("data")
outsidechan:push("test")
outsidechan:push({1, 2})
```

jgmdev commented:
Here some reference videos of the speed difference of coroutines vs threads using the [findfileimproved](https://github.com/lite-xl/lite-xl-plugins/pull/29) plugin:

Coroutines:

https://user-images.githubusercontent.com/1702572/153547542-f659c559-6bb8-42b5-bdd4-01e1e1f75072.mp4

Threads:

https://user-images.githubusercontent.com/1702572/153547603-af26bd9d-a8f3-4447-91ec-00a089cc8391.mp4

adamharrison commented:
To put my thoughts down officially, and not just in discord chat:

I like it. It's true for larger project, this is a major concern. However, I am not generally in favour of adding another method of compiling optional native modules; I think if we're going to do this, we should subsume this into the native plugin functionality. (Although, I am potentially open to make this a core plugin we distribute with lite like the other core plugins). 

Given that you're already using the native plugin API, that should work out totally fine; but basically let's structure this under the plugins directory, what I'm thinking we do is have a folder for the plugin (like some plugins already have). Everything else looks pretty good, though.

jgmdev commented:
> what I'm thinking we do is have a folder for the plugin (like some plugins already have), and just put all the stuff currently in the modules directory in there.

What directory would that be? Also I have to fix the memory leak in `channel:pop()` inherited from the luasdl2 codebase, I may restructure the queue/list algorithm for that.

adamharrison commented:
I mean `data/plugins/threads`, for example.

jgmdev commented:
> I mean `data/plugins/threads`, for example.

Ohh I see, but wouldn't a path like that be confusing with lua code? Shoudn't the path be more clear to differentiate Lua plugins and C plugins?

jgmdev commented:
In the meantime got the channel memory leak and double free segmentation fault sorted out.

adamharrison commented:
> [](https://github.com/jgmdev)Ohh I see, but wouldn't a path like that be confusing with lua code? Shoudn't the path be more clear to differentiate Lua plugins and C plugins?

Maybe. It's a good question. As it stands, the normal way to do native plugins has been to treat them like any other plugins; following the lua paradigm. However, if we want to make a distinction, we could. I'm not sure we *should*, mind you; generally it's pretty clear if a plugin needs to be compiled or not, but we don't exactly have many public native plugins, so we could change this.

jgmdev commented:
Eliminated the `modules/` dir and moved its content to `data/plugins/` and made necessary changes to meson.build to handle this.

jgmdev commented:
Added execution of core/start.lua on a thread own state to allow requiring from the path's where lite-xl stuff is saved, so threads can now require other plugins/libs or core code like core.common which has nice utility functions.

jgmdev commented:
I think I finished squashing the remaining bugs I found while implementing a multi-threaded file search, below the comparison of current implementation and new threaded implementation:

**Actual implementation on Lite-XL**

https://user-images.githubusercontent.com/1702572/157343664-7d280418-4bf8-47c0-8fa1-58c86e80ddfb.mp4

**New Multi-threaded implementation**

https://user-images.githubusercontent.com/1702572/157343773-d2442bbe-054c-4cf8-b5e4-365e080f261e.mp4


And here is the modified [projectsearch.lua](https://gist.github.com/jgmdev/22a7f63a16726d1372e14880573f9bab) that optionally supports threading for anyone to try.

jgmdev commented:
Reporting that with #880 (jit enabled) the results reported above improved especially for the threaded one.

For the co-routine one went from: ~10s to ~7s
For the multi-threaded went from: ~5s to ~2s

Seems that jit improves the coroutine performance nicely.

jgmdev commented:
Updated testing threaded projectsearch plugin to properly work with latest master changes:

https://gist.github.com/jgmdev/22a7f63a16726d1372e14880573f9bab

adamharrison commented:
Tenatively closed for now, this functionality is available either as a library in https://github.com/jgmdev/lite-xl-threads or in Pragtical: https://github.com/pragtical/pragtical.
